hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/DualFileWriter.java

```diff
@@ -0,0 +1,255 @@
+    this.newVersionBehavior = newVersionBehavior;
+    writers = new ArrayList<>(2);
+    initRowState();
+  }
+
+  private void initRowState() {
+    deleteFamily = null;
+    deleteFamilyVersionList.clear();
+    lastCell = null;
+  }
+
+  private void initColumnState() {
+    livePutCellCount = 0;
+    deleteColumn = null;
+    deleteColumnVersionList.clear();
+
+  }
+
+  private void addLiveVersion(Cell cell) throws IOException {
+    if (liveVersionWriter == null) {
+      liveVersionWriter = writerFactory.createWriter();
+      writers.add(liveVersionWriter);
+    }
+    liveVersionWriter.append(cell);
+  }
+
+  private void addHistoricalVersion(Cell cell) throws IOException {
+    if (historicalVersionWriter == null) {
+      historicalVersionWriter = writerFactory.createWriter();
+      writers.add(historicalVersionWriter);
+    }
+    historicalVersionWriter.append(cell);
+  }
+
+  private boolean isDeletedByDeleteFamily(Cell cell) {
+    return deleteFamily != null && (deleteFamily.getTimestamp() > cell.getTimestamp()
+      || (deleteFamily.getTimestamp() == cell.getTimestamp()
+        && (!newVersionBehavior || cell.getSequenceId() < deleteFamily.getSequenceId())));
+  }
+
+  private boolean isDeletedByDeleteFamilyVersion(Cell cell) {
+    for (Cell deleteFamilyVersion : deleteFamilyVersionList) {
+      if (
+        deleteFamilyVersion.getTimestamp() == cell.getTimestamp()
+          && (!newVersionBehavior || cell.getSequenceId() < deleteFamilyVersion.getSequenceId())
+      ) return true;
+    }
+    return false;
+  }
+
+  private boolean isDeletedByDeleteColumn(Cell cell) {
+    return deleteColumn != null && (deleteColumn.getTimestamp() > cell.getTimestamp()
+      || (deleteColumn.getTimestamp() == cell.getTimestamp()
+        && (!newVersionBehavior || cell.getSequenceId() < deleteColumn.getSequenceId())));
+  }
+
+  private boolean isDeletedByDeleteColumnVersion(Cell cell) {
+    for (Cell deleteColumnVersion : deleteColumnVersionList) {
+      if (
+        deleteColumnVersion.getTimestamp() == cell.getTimestamp()
+          && (!newVersionBehavior || cell.getSequenceId() < deleteColumnVersion.getSequenceId())
+      ) return true;
+    }
+    return false;
+  }
+
+  private boolean isDeleted(Cell cell) {
+    return isDeletedByDeleteFamily(cell) || isDeletedByDeleteColumn(cell)
+      || isDeletedByDeleteFamilyVersion(cell) || isDeletedByDeleteColumnVersion(cell);
+  }
+
+  private void appendCell(Cell cell) throws IOException {
+    if ((lastCell == null || !CellUtil.matchingColumn(lastCell, cell))) {
+      initColumnState();
+    }
+    if (cell.getType() == Cell.Type.DeleteFamily) {
+      if (deleteFamily == null) {
+        if (cell.getType() == Cell.Type.DeleteFamily) {
+          deleteFamily = cell;
+          addLiveVersion(cell);
+        } else {
+          addHistoricalVersion(cell);
+        }
+      }
+    } else if (cell.getType() == Cell.Type.DeleteFamilyVersion) {
+      if (deleteFamily == null) {
```

Comment:
Do we not need check here to see if the DeleteFamilyVersion is deleted by DeleteFamily?

Reply:
Good catch! Yes, we need to do that.